modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/AbstractSetOpNode.java

```diff
@@ -0,0 +1,366 @@
+    /** */
+    private final AggregateType type;
+
+    /** */
+    private final Grouping<Row> grouping;
+
+    /** */
+    private int requested;
+
+    /** */
+    private int waiting;
+
+    /** Current source index. */
+    private int curSrcIdx;
+
+    /** */
+    private boolean inLoop;
+
+    /** */
+    protected AbstractSetOpNode(ExecutionContext<Row> ctx, RelDataType rowType, AggregateType type, boolean all,
+        RowFactory<Row> rowFactory, Grouping<Row> grouping) {
+        super(ctx, rowType);
+
+        this.type = type;
+        this.grouping = grouping;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void request(int rowsCnt) throws Exception {
+        assert !F.isEmpty(sources());
+        assert rowsCnt > 0 && requested == 0;
+        assert waiting <= 0;
+
+        checkState();
+
+        requested = rowsCnt;
+
+        if (waiting == 0)
+            sources().get(curSrcIdx).request(waiting = IN_BUFFER_SIZE);
+        else if (!inLoop)
+            context().execute(this::flush, this::onError);
+    }
+
+    /** */
+    public void push(Row row, int idx) throws Exception {
+        assert downstream() != null;
+        assert waiting > 0;
+
+        checkState();
+
+        waiting--;
+
+        grouping.add(row, idx);
+
+        if (waiting == 0)
+            sources().get(curSrcIdx).request(waiting = IN_BUFFER_SIZE);
+    }
+
+    /** */
+    public void end(int idx) throws Exception {
+        assert downstream() != null;
+        assert waiting > 0;
+        assert curSrcIdx == idx;
+
+        checkState();
+
+        grouping.endOfSet(idx);
+
+        if (type == AggregateType.SINGLE && grouping.isEmpty())
+            curSrcIdx = sources().size(); // Skip subsequent sources.
+        else
+            curSrcIdx++;
+
+        if (curSrcIdx >= sources().size()) {
+            waiting = -1;
+
+            flush();
+        }
+        else
+            sources().get(curSrcIdx).request(waiting);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void rewindInternal() {
+        requested = 0;
+        waiting = 0;
+        grouping.groups.clear();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected Downstream<Row> requestDownstream(int idx) {
+        return new Downstream<Row>() {
+            @Override public void push(Row row) throws Exception {
+                AbstractSetOpNode.this.push(row, idx);
+            }
+
+            @Override public void end() throws Exception {
+                AbstractSetOpNode.this.end(idx);
+            }
+
+            @Override public void onError(Throwable e) {
+                AbstractSetOpNode.this.onError(e);
+            }
+        };
+    }
+
+    /** */
+    private void flush() throws Exception {
+        if (isClosed())
+            return;
+
+        checkState();
+
+        assert waiting == -1;
+
+        int processed = 0;
+
+        inLoop = true;
+
+        try {
+            while (requested > 0 && !grouping.isEmpty()) {
```

Comment:
Is it possible to iterate over this loop more than once?

Reply:
Looks like only one iteration is possible. Changed to `if`